cpp/data_structures/dynamic_tree/DynamicTree.hpp

```diff
@@ -0,0 +1,441 @@
+ private:
+  // main node in DynamicTree
+  std::shared_ptr<DTNode<T>> root;
+
+  // o(log n)
+  // search for children node in current node
+  bool binary_search(std::string const& target_name, int& target_index) {
+    int min = 0;
+    int max = current_node->children.size() - 1;
+    int mid = (min + max) / 2;
+
+    while (min <= max) {
+      // in case target found
+      if (current_node->children[mid]->name == target_name) {
+        if (&target_index != NULL) {
+          target_index = mid;
+        }
+        return true;
+      }
+
+      // going left
+      if (current_node->children[mid]->name.compare(target_name) > 0) {
+        max = (mid - 1);
+      }
+
+      // going right
+      if (current_node->children[mid]->name.compare(target_name) < 0) {
+        min = (mid + 1);
+      }
+
+      // update mid
+      mid = (min + max) / 2;
+    }
+
+    // in case not found
+    if (&target_index != NULL) {
+      target_index = -1;
+    }
+    return false;
+  }
+
+  // o(log n)
+  // search for children in temp node
+  int binary_search(std::string const& target_name,
+                    std::shared_ptr<DTNode<T>> temp) {
+    int min = 0;
+    int max = temp->children.size() - 1;
+    int mid = (min + max) / 2;
+
+    while (min <= max) {
+      // in case target found
+      if (temp->children[mid]->name == target_name) {
+        return mid;
+      }
+
+      // going left
+      if (temp->children[mid]->name.compare(target_name) > 0) {
+        max = (mid - 1);
+      }
+
+      // going right
+      if (temp->children[mid]->name.compare(target_name) < 0) {
+        min = (mid + 1);
+      }
+
+      // update mid
+      mid = (min + max) / 2;
+    }
+
+    // in case not found
+    return -1;
+  }
+
+  // o(1)
+  // function take a "temp node" & make a simple check if that temp include a
+  // parent with specific name return => "parent pointer" or NULL
+  std::shared_ptr<DTNode<T>> search_up(std::string& parent_name,
+                                       std::shared_ptr<DTNode<T>> temp) {
+    if (temp->parent != NULL && temp->parent->name == parent_name) {
+      return temp->parent;
+    } else
+      return NULL;
+  }
+
+  // o(1) --> o(log n)
+  // like "search_up" function , but this one for looking "down" between
+  // children
+  std::shared_ptr<DTNode<T>> search_down(std::string& child_name,
+                                         std::shared_ptr<DTNode<T>> temp) {
+    // o(log n)
+    // search for target
+    int index = this->binary_search(child_name, temp);
+
+    if (index != -1) {
+      return temp->children[index];
+    }
+
+    return NULL;
+  }
+
+  // o(path) --> o(log n * path)
+  /*
+      like go_to but this require a hole path of "names" in one direction
+      note !! travel will happend only if "the whole path of names" is valid
+      otherwise nothing will be happen & return will be false
+  */
+  bool travel(std::vector<std::string> const& path_of_names, bool up = false) {
```

Comment:
```suggestion
  bool traverse(std::vector<std::string> const& path_of_names, bool up = false) {
```

Mayeb traverse is a better name?